Add DevServices support for Vault extension

kdubb opened the pull request:
Enables DevServices for Vault.

* A single Vault container is supported (equivalent to the extension)
* The configuration allows for enabling engines:
* * Transit
* * PKI

Note that due to my unfamiliarity with how DevServices worked (although it seems mostly clear now) this is implemented essentially a copy of the recent PR #17070 for Redis. If there is complexity that can be removed or the feature can be better organized I'm all ears.

kdubb commented:
@vsevel Please review!

netodevel reviewed (changes requested):
I made some comments.
And I missed integration tests.

kdubb commented on extensions/vault/deployment/src/main/java/io/quarkus/vault/deployment/DevServicesVaultProcessor.java:
```diff
@@ -0,0 +1,202 @@
+        DevServicesConfig currentDevServicesConfiguration = config.devservices;
+
+        // figure out if we need to shut down and restart any existing Vault container
+        // if not and the Vault container have already started we just return
```
This only restarts the Vault container if the user changed the devservices configuration and is currently running in dev mode or continuous testing.

kdubb commented on extensions/vault/deployment/src/main/java/io/quarkus/vault/deployment/DevServicesVaultProcessor.java:
```diff
@@ -0,0 +1,202 @@
+            Thread closeHookThread = new Thread(closeTask, "Vault container shutdown thread");
+            Runtime.getRuntime().addShutdownHook(closeHookThread);
+            ((QuarkusClassLoader) cl.parent()).addCloseTask(new Runnable() {
+                @Override
```
As I said before, I followed the model of #17070; which has been merged this morning. Also, I'm unaware of a shutdown build item that can be used for this purpose

NetoDevel replied:
@kdubb, all right.

kdubb commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/DevServicesConfig.java:
```diff
@@ -0,0 +1,76 @@
+import io.quarkus.runtime.annotations.ConfigItem;
+
+@ConfigGroup
+public class DevServicesConfig {
```
`VaultBuildTimeConfig` is the root and already contains the mentioned phase.

kdubb committed:
Add DevServices support for Vault extension

kdubb commented:
@machi1990 rebased

vsevel commented on extensions/vault/deployment/src/main/java/io/quarkus/vault/deployment/DevServicesVaultProcessor.java:
```diff
@@ -0,0 +1,202 @@
+            BuildProducer<RunTimeConfigurationDefaultBuildItem> runTimeConfiguration,
+            BuildProducer<ServiceStartBuildItem> serviceStartBuildItemBuildProducer, VaultBuildTimeConfig config) {
+
+        DevServicesConfig currentDevServicesConfiguration = config.devservices;
```
may be we should be consistent on naming: `devServices` instead of `devservices`

kdubb replied:
The configuration key `devservices` (all lowercase) is consistent with all the other DevServices configuration keys (e.g. Datasources, Mongo, Redis, etc.).

vsevel commented on extensions/vault/deployment/src/main/java/io/quarkus/vault/deployment/DevServicesVaultProcessor.java:
```diff
@@ -0,0 +1,202 @@
+
+public class DevServicesVaultProcessor {
+    private static final Logger log = Logger.getLogger(DevServicesVaultProcessor.class);
+    private static final String VAULT_IMAGE = "vault:1.7.1";
```
we should be consistent with `VaultTestExtension.getVaultImage()`. Let's upgrade everything to `1.7.1`.

kdubb replied:
Added a `VaultVersions` class with the constant `VAULT_TEST_VERSION` in the `quarkus-vault` module. Both `DeviceServicesVaultProcessor` and `VaultTestExtensions` now select their version based on that constant.

vsevel commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/DevServicesConfig.java:
```diff
@@ -0,0 +1,76 @@
+     * by default, unless there is an existing configuration present.
+     * <p>
+     * When DevServices is enabled Quarkus will attempt to automatically configure and start
+     * a database when running in Dev or Test mode and when Docker is running.
```
_vault instance_ instead of _database_

vsevel reviewed (dismissed):
not familiar with dev services. I have only minor points.

vsevel commented:
I think it is worth adding a small paragraph in the vault's documentation.

kdubb commented:
@vsevel I've updated the docs reference DevServices support and configuration. Also squashed and rebased to top of main.

You said you are not familiar with DevServices. Who should be merging this?

kdubb commented:
@gsmet You reviewed and merged the Redis DevServices change, can you take a look at this as well?

gsmet commented on docs/src/main/asciidoc/vault.adoc:
```diff
@@ -278,6 +283,11 @@ quarkus.vault.authentication.userpass.password=sinclair
 
+[TIP]
+====
+Using <<dev-services,DevServices>>, Quarkus will take care of configuring the Vault connection in dev & test mode!
```
```suggestion
Using <<dev-services,DevServices>>, Quarkus will take care of configuring the Vault connection in dev and test mode.
```

gsmet commented on docs/src/main/asciidoc/vault.adoc:
```diff
@@ -508,6 +518,42 @@ path "auth/kubernetes/role/*" {
+Quarkus will automatically start a Vault container when running tests or in dev-mode, and automatically
+configure the connection.
+
+When running the production version of the application, the Vault connection need to be configured as normal,
```
```suggestion
When running the production version of the application, the Vault connection needs to be configured as normal,
```

gsmet commented on docs/src/main/asciidoc/vault.adoc:
```diff
@@ -508,6 +518,42 @@ path "auth/kubernetes/role/*" {
+
+=== Automatic Secret Engine Provisioning
+
+To help with provisioning the automatically managed Vault instance you can enable certain secret engines.
```
```suggestion
To help with provisioning the automatically managed Vault instance, you can enable certain secret engines.
```

gsmet reviewed (changes requested):
I spotted a couple of issues in the doc, could you have another look? Apart from that, it looks good to go, thanks.

gsmet commented:
@vsevel could you check your comments are addressed? If so, it looks good to go.

vsevel commented:
> @vsevel could you check your comments are addressed? If so, it looks good to go.

hi @gsmet. yes, they have been.
thanks @kdubb for this contribution.
